src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,675 @@
+                                      packet + sizeof(eth_ip_udp_t) );
+
+  ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+  ulong sig = fd_disco_netmux_sig( ip, port, FD_NETMUX_SIG_MIN_HDR_SZ, DST_PROTO_OUTGOING, (ushort)0 );
+  fd_mcache_publish( ctx->net_out_mcache, ctx->net_out_depth, ctx->net_out_seq, sig, ctx->net_out_chunk, packet_sz, 0UL, tsorig, tspub );
+  ctx->net_out_seq   = fd_seq_inc( ctx->net_out_seq, 1UL );
+  ctx->net_out_chunk = fd_dcache_compact_next( ctx->net_out_chunk, packet_sz, ctx->net_out_chunk0, ctx->net_out_wmark );
+}
+
+static inline void
+handle_new_cluster_contact_info( fd_repair_tile_ctx_t * ctx,
+                                 uchar const    * buf ) {
+  ulong const * header = (ulong const *)fd_type_pun_const( buf );
+
+  ulong dest_cnt = header[ 0 ];
+
+  if( dest_cnt >= MAX_REPAIR_PEERS ) {
+    FD_LOG_ERR(( "Cluster nodes had %lu destinations, which was more than the max of %lu", dest_cnt, MAX_REPAIR_PEERS ));
+  }
+
+  fd_shred_dest_wire_t const * in_dests = fd_type_pun_const( header+1UL );
+
+  for( ulong i=0UL; i<dest_cnt; i++ ) {
+    fd_repair_peer_addr_t repair_peer = {
+      .addr = in_dests[i].ip4_addr,
+      .port = in_dests[i].udp_port,
+    };
+   
+    fd_repair_add_active_peer( ctx->repair, &repair_peer, in_dests[i].pubkey );
+  }
+}
+
+static inline void
+handle_new_repair_requests( fd_repair_tile_ctx_t * ctx,
+                            uchar const    * buf,
+                            ulong buf_sz ) {
+
+  fd_repair_request_t const * repair_reqs = (fd_repair_request_t const *)fd_type_pun_const( buf );
+  ulong repair_req_cnt = buf_sz;
+  // FD_LOG_WARNING(("Repair requests %lu received %lu", repair_req_cnt, buf_sz));
+  for( ulong i = 0; i < repair_req_cnt; i++ ) {
+    fd_repair_request_t const * repair_req = &repair_reqs[i];
+    int rc = 0;
+    switch(repair_req->type) {
+      case FD_REPAIR_REQ_TYPE_NEED_WINDOW_INDEX: {
+        rc = fd_repair_need_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_HIGHEST_WINDOW_INDEX: {
+        rc = fd_repair_need_highest_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_ORPHAN: {
+        rc = fd_repair_need_orphan( ctx->repair, repair_req->slot );
+        break;
+      }
+    }
+    
+    if( rc < 0 ) {
+      FD_LOG_WARNING(( "failed to issue repair request" ));
+    }
+  }
+
+}
+
+static inline void
+handle_new_stake_weights( fd_repair_tile_ctx_t * ctx,
+                          uchar const    * buf ) {
+  ulong const * header = (ulong const *)fd_type_pun_const( buf );
+
+  ulong stakes_cnt = header[ 1 ]; 
+
+  if( stakes_cnt >= MAX_REPAIR_PEERS ) {
+    FD_LOG_ERR(( "Cluster nodes had %lu stake weights, which was more than the max of %lu", stakes_cnt, MAX_REPAIR_PEERS ));
+  }
+
+  fd_stake_weight_t const * in_stake_weights = fd_type_pun_const( header+4UL );
```

Comment:
Actually this stuff is kind of suspect. You should probably just define a type for these overall, eg,

fd_stake_weight_t {
    ulong cnt;
    .... weights;
}

and cast (no need to type pun). You also probably want to validate the sz and not read out of bounds? You can see how the existing tiles do this for the right style, for all of the above parsing functions.

Reply:
shifted to using stake ci